README.markdown
xen-tools
=========

* [Homepage](http://www.xen-tools.org/software/xen-tools)
* Git Repository:
  * [at Gitorious](http://gitorious.org/xen-tools)
  * [at GitHub](http://github.com/xtaran/xen-tools)
* [Mailing Lists](http://www.xen-tools.org/software/xen-tools/lists.html)

About
-----

xen-tools is a collection of simple shell , perl, python scripts which allow you to easily 
create new Xen guest domains upon your NATIVE Xen host server.

xen-tools gives you the abillty fully control what and how "Items" get auto-deployed with the supported install methodes  
while  working with Xen  guest images under Linux .

xen-tools  Support as well FAI , pupped , sudo and  some more .

xen-tools are higly configurable and simply to addapt to your own needs .


Using this software, you can easily create new
[Xen](http://www.xen.org) guests configured to be accessible over the
network via [OpenSSH](http://www.openssh.org/).

xen-tools currently has scripts to install most releases of
[Debian](http://www.debian.org/) (starting with 3.1 "Sarge") and
[Ubuntu](http://www.ubuntu.com/) (starting with 6.06 LTS "Dapper") and
some RPM-based distributions. On the Dom0 side all current Xen
supporting distributions should work.


Supported Distributions
-----

The Folloing  releases are tested and known to work reliably  aka "Supported" :


### Debian

* Sarge 3.1 (i386 and DomU only)
* Etch 4.0 (Dom0 no more tested)
* Lenny 5.0
* Squeeze 6.0
* Wheezy 7
* Jessie 8 (under development)
* Sid (always under development; works at least at the moment of writing :-)

### Ubuntu

(only DomUs tested)

* Dapper Drake 6.06 (needs editing of `/usr/share/debootstrap/scripts/dapper` since at least debootstrap version 1.0.37, see [#659360][1])
* Edgy Eft 6.10 (needs editing of `/usr/share/debootstrap/scripts/edgy` since at least debootstrap version 1.0.37, see [#659360][1])
* Feisty Fawn 7.04
* Gutsy Gibbon 7.10
* Hardy Heron 8.04 (see [Installing Ubuntu 8.04 as DomU][2])
* Interpid Ibex 8.10
* Jaunty Jackaplope 9.04
* Karmic Koala 9.10
* Lucid Lynx 10.04
* Maverick Meerkat 10.10
* Natty Narwhal 11.04
* Oneiric Ocelot 11.10
* Precise Pangolin 12.04
* Quantal Quetzal 12.10
* Raring Ringtail 13.04
* Saucy Salamander 13.10
* Trusty Tahr 14.04 (LTS, under development)

[1]: http://bugs.debian.org/659360
    "debootstrap in Wheezy can no more build Ubuntu Edgy or earlier"

[2]: http://www.linux-vserver.org/Installing_Ubuntu_8.04_Hardy_as_guest
    "There is an issue with debootstrap on hardy not installing ksyslogd."

### CentOS

only DomUs tested, pygrub support incomplete on older then 6.0 .

* CentOS 5
* CentOS 6

However, currently only centos 6.5 and up releases are fully tested and known to work reliably with pygrub
due now kernel 3.x used with dracut.

* CentOS 6.5
* CentOS 7.0 - work in progress target -> q3 or q4 2014 

### OpenSuSE / SuSE / SLES

* OSS 13.1


Packages
--------

xen-tools are available prepackaged in Debian (and derivates) and as
source tar-ball for local installation. Installing from source should
work flawlessly on most Linux systems that meet the installation
requirements.

Requirements
------------

To use these tools you'll need the following software on Dom0: <br>

   Tool                 Source URL                        - what      + distro  <br>
    
* [debootstrap](http://packages.debian.org/debootstrap)   - bootstrap +  debian/ubuntu  <br>
* [fakeroot](http://packages.debian.org/fakeroot)         - chroot helper <br>
* [rpmstrap](https://github.com/blipvert/rpmstrap)        - rpmstrap   +  rpmbased install helper <br>
* [rinse] (http://collab-maint.alioth.debian.org/rinse)    - bootstrap  +   RHEL / FC / SuSE / SLES  <br>

* Perl and the following Perl modules
  Requires: perl(Text::Template) <br>
  Requires: perl(Config::IniFiles) <br>
  Requires: perl(Expect::Simple) perl(Expect) <br>
  Requires: perl(Getopt::Long) <br>
  Requires: perl(LWP::UserAgent) <br>
  Requires: perl(Pod::Usage) <br>
  Requires: perl(File::Basename) <br>
  Requires: perl(File::Find) <br>
  Requires: perl(File::Path) <br>
  Requires: perl(File::Temp) <br>
  Requires: perl(File::Which) <br>
  Requires: perl(File::Copy) <br>
  Requires: perl(Term::UI) <br>
  Requires: perl(Term::ReadLine) <br>
  Requires: perl(IPC::Cmd) <br>
  Requires: perl(Digest::MD5) <br>
  Requires: perl(Data::Validate::URI) <br>
  Requires: perl(Data::Validate::IP) <br>
  Requires: perl(Data::Validate::Domain) <br>

Most of them are delivered by distributions , thus this are only an question of allmost  <br>
to use the native distro package tools (apt-get / yum / zypper / yast / rpm / ... )  <br>
If not found in distribution - use i.e http://www.rpmfind.net, http://www.debfind.de <br>
Or install them via  perl CPAN . <br>
 
* To use the --hash_method=expect you need expect installed on domU: <br>
  [expect] (http://expect.sourceforge.net/) <br>
  expect  require tcl to be installed. <br>
  
This been currently in use at ( oss13.1 : 29-setup-expect ) for autpdeploy inital random root pwd at opensuse 13.x  xen guest. <br>
expect can be autoremoved as well after with --finalrole=remove_rpm_expect for those that have an security concern about it . <br>

* gnu make,and compiler ,  rpmbuild tools  if you are not installing through a package manager.

* [rinse] (http://collab-maint.alioth.debian.org/rinse)  
Sample configs for centos6.5 , opensuse 13.1 added under examples - they are fully tested .<br>
Rinse require fakeroot and rpmstrap to work cleanly. <br>

If you wish to create new Xen instances which may be controlled by <br>
users via a login shell you can have a look at the (currently <br>
unmaintained) [xen-shell](http://xen-tools.org/software/xen-shell/)
project.


Installation from Source
------------

As root or with sudo, execute `make instal`.

For Debian/Ubuntu  see `debian/README.source` how to build the Debian package from a<br>
checked out copy of the git repository (i.e. without a source tarball).<br>


Installation from RPMbuild
------------

 - picup specfile & build  patches  from  https://github.com/remsnet/xen-tools-rpmbuild
   add them to rpmbuild´s SPECS dir  i.e /usr/src/packages/SPECS

 - picup xen-tools sources from this repro, use "master"  ,     rpmbuild -ba <spec> ,     rpm -i <rpm>

 
The Scripts
-----------

Here is a brief description of each included script, for more thorough
documentation please consult the appropriate man page.


### xen-create-image

This script is designed to create new images which may be used
with the Xen hypervisor.

This script performs the initial setup, then delegates the real
work to a collection of helper scripts:

* `xt-install-image`: Installs a distribution into a directory.

* `xt-customize-image`: Run a collection of hook scripts to configure
  the freshly installed system.

* `xt-create-xen-config`: Create a configuration file in `/etc/xen`
  such that Xen can boot the newly created machine.

* `xt-guess-suite-and-mirror`: In case of a Debian or Ubuntu Dom0,
  this script tries to guess the most suitable suite and mirror for
  DomUs based on the Dom0's `/etc/apt/sources.list`.


### xen-create-nfs

This script is similar in spirit to `xen-create-image`, but much less
complex.  It allows the creation of Xen guests which are diskless,
mounting their root filesystem over a remote NFS-share.

There are not many options to tweak, but still a useful addition 
to the suite.


### xen-delete-image

This script will allow you to completely remove Xen instances which
have previously been created by `xen-create-image`, this includes
removing the storage block devices from the system, and deleting the
Xen configuration file.


### xen-list-images

List all the created images beneath a given root directory along with
a brief overview of their setup details.


### xen-update-image

This script runs "apt-get update; apt-get upgrade" for a given Xen
image.

#### NOTES

* The image should not be running or corruption will occur!
* The script should only be used for Xen instances of Debian or a
  Debian-derived distribution.


-- 
The Xen-Tools Developer Team


NEWS.markdown
xen-tools 4.4.rcis2-dev03 (Forked release  12-July-2014)
- added  patch-fix --hash_method=expect as an bugfix workarround for the incompatible sha512 salt for opensuse / sles
  this expects  dir  /etc/xen-tools/expect exist and look for i.e an expect chat script opensuse-13.1-passwd.exp 
  This require perl modules  IPC::Cmd  and File::Copy on dom0 and expect to be installed in the domU.

- updated & moved to grub2  for opensuse13.1 hooks 23-setup-grub, 26-setup-kernel. <br>
  Thanks to Ismael Luceno for the patch Contribution.<br>
  This require copy of the pygrub  PYGRUB_VER = 0.6 , and an backport copy from xen 4.4 of the GrubConf.py <br>
  least to work & function.

xen-tools 4.4.rcis2-dev02 (Forked release  22-June-2014)
- added 24-setup-initramfs,
  busybox initramfs for opensuse 13.1 as an workaround for issues not mounted /dev/xvda2 , 
   caused by not automounted devtmpfs witch caused that the xenblk (xen bl front ) was not loated .
  oss 13.1 dracut BUG : don not generate an valid initramfs 
- oss 13.1 boots now  as centos6.5 with xen 4.4 domU - currenty with network issue.
- added oss13.1 hook 23-setup-udev-rules
- added oss13.1 hook 23-setup-grub 
- added oss13.1 hook 21-setup-keyboard
- added oss13.1 hook 22-setup-bashrc
- updated 95-clean-image, 90-make-fstab, 55-create-dev , 40-setup-networking

xen-tools 4.4.rcis2-dev01 (Forked release  14-June-2014)
- added experimental opensuse 13.1 hooek support.
- added prestage funtion for distros like opensuse to setup cleanly chroot - without  that  , the OSS autodeplyment failed .
- patched hardcoded /bin/ls /bin/cp with more sufficent : check /usr/bin and fllow and check symlikns
  This are required due LSB standard 2.3 / 3.0dev
- xen-tools & rinse Perl patches & fixes by Adrian Arnautu /Romania, Bucharest   arnautu dot adrian at gmail dot com

xen-tools 4.4.rcis1 (Forked release  27-Mai-2014)
======================================
 - Steve left, handed over to xen-tools.org Folks .
 - New xen-tools Maintainer : Alex Beckert.
 - xen-tools forked from xen-tools org in April 2014 for Remsnet LTD support patches 
 - centos 6.5 support added
 - centos 7.0 support added ( work in progress )
 - Opensuse 13.1 support added ( work in progress )
 - Fedora 19/20 support added ( work in progress )
 - updated rinse 
 - created rpm specs  for all depend rpm based deployments see https://github.com/remsnet/xen-tools-rpmbuild 

 Steven kemp Noted April 2014 to us :
 "Thanks for sharing, but you should know that I no longer take
  part in any xen-tools development. Instead the project was
  handed over to xen-tools.org folks who had more time and energy to continue  with it.
  Your contribution would be best shared with the mailing list here:
  http://xen-tools.org/mailman/listinfo/xen-tools-discuss
  Apologies for the slow reply; I've been on holiday.
  Steve
  -- 
  http://www.steve.org.uk/ "


xen-tools 4.3.1 (released 30-Jun-2012)
======================================

Bugfix Release only


xen-tools 4.3 (released 26-Jun-2012)
====================================

Bug Fixes
---------

* Fix several testuite failures depending on the build host's
  installation.

Other Changes
-------------

* Remove most Mercurial traces


xen-tools 4.3rc1 (released 08 Jun 2012)
=======================================

New Features and Major Changes
------------------------------

* Massive code deduplication in hooks directory

New Options
-----------

    --dontformat (xen-create-image)
    --finalrole  (xen-create-image)
    --apt_proxy  (xen-create-image)

Newly Supported Distribution Releases
-------------------------------------

* Ubuntu 11.10 Oneiric
* Ubuntu 12.04 Precise
* Ubuntu 12.10 Quantal
* CentOS 6


xen-tools 4.2.1 (released 17 Mar 2011)
======================================

Bugfix Release only


xen-tools 4.2 (released 05 Oct 2010)
====================================

First final release of the new Xen-Tools Team.

Supports Ubuntu up to 11.04 (Natty) and Debian up to 7.0 (Wheezy).


New Options
-----------

    --debootstrap-cmd (xen-create-image and xt-install-image)

New Features and Major Changes
------------------------------

* Uses hvc0 and xvda devices by default
* Also supports cdebootstrap
* Preliminary btrfs support.
* Uses GeoIP for Debian mirrors: Default Debian mirror is now
  cdn.debian.net, see http://wiki.debian.org/DebianGeoMirror for
  details.
* New helper program xt-guess-suite-and-mirror, used to find the
  default mirror and suite.


examples/rinse/deploy-VM-with-lvm-test1.sh
#!/bin/sh

IP_ADDR=10.20.0.10
NET_MASK=255.255.255.0
GATEWAY=10.20.0.1
HOSTNAME=test1
DNS1=10.20.0.1

test -d /var/tmp/xen-create-image  || mkdir -p /var/tmp/xen-create-image

> /var/tmp/xen-create-image/networkcentos.tmpl
echo IP_ADDR=10.20.0.10 >>/var/tmp/xen-create-image/networkcentos.tmpl
echo NET_MASK=255.255.255.0 >>/var/tmp/xen-create-image/networkcentos.tmpl
echo GATE_WAY=10.20.0.1 >>/var/tmp/xen-create-image/networkcentos.tmpl
echo HOST_NAME=test1 >>/var/tmp/xen-create-image/networkcentos.tmpl
echo NAME_SERVER1=10.20.0.1  >>/var/tmp/xen-create-image/networkcentos.tmpl
echo NAME_SERVER2=10.20.0.2  >>/var/tmp/xen-create-image/networkcentos.tmpl


xen-create-image  --hostname=$HOSTNAME \
        --image=full \
        --lvm=vg2 \
        --size=10GB --fs=ext3 \
        --swap=2G \
        --disk_device=xvda \
        --role=editor \
        --roledir=/etc/xen-tools/role.d \
        --hooks=1 \
        --pygrub \
        --nodhcp \
        --force \
        --vcpus=2 \
        --memory=512M \
        --bridge=br1 \
        --gateway=$GATEWAY \
        --ip=$IP_ADDR \
        --netmask=$NET_MASK \
        --install-method=rinse \
        --dist=centos-6.5 \
        --mirror=http://<your-mirror.centos.org>/centos/6.5/os/x86_64 \
        --arch=amd64 \
        --kernel=/bootxen/vmlinuz-xen6.5-install \
        --initrd=/bootxen/initrd-xen6.5-install


examples/rinse/usr/lib/rinse/opensuse-13.1/pre-install.sh
#!/bin/sh
#
#  Customise the distribution pre-install.
#

prefix=$1

if [ ! -d "${prefix}" ]; then
  echo "Serious error - the named directory doesn't exist."
  exit
fi


#
#  3.  Ensure that zypper has a working configuration file.
#
arch=i386
if [ $ARCH = "amd64" ] ; then
    arch=x86_64
fi

echo "  Creating zypper repo entry"
[ -d "${prefix}/etc/zypp/repos.d" ] || mkdir -p ${prefix}/etc/zypp/repos.d
cat > ${prefix}/etc/zypp/repos.d/${dist}.repo <<EOF
[opensuse]
name=${dist}
baseurl=$(dirname $(dirname ${mirror}))
enabled=1
gpgcheck=1

EOF


#
#  4.  Run pre install  aaa_base - we require the base bevor we run post
#  ->moved to prestage/opensuse-13.1.sh


#
#  5.  Clean up
#
echo "  Cleaning up"
chroot ${prefix} /usr/bin/zypper clean
chroot ${prefix} /usr/bin/zypper ref
umount ${prefix}/proc
umount ${prefix}/sys


examples/rinse/usr/lib/rinse/opensuse-13.1/post-install.sh
#!/bin/sh
#
#  Customise the distribution post-install.
#

prefix=$1

if [ ! -d "${prefix}" ]; then
  echo "Serious error - the named directory doesn't exist."
  exit
fi

#
#  2.  Copy the cached .RPM files into the zypper directory, so that
#     zypper doesn't need to fetch them again.
#
echo "  Setting up zypper cache"

if [ ! -d "${prefix}/var/cache/zypp/packages/opensuse/suse/${arch}" ]; then
    mkdir -p ${prefix}/var/cache/zypp/packages/opensuse/suse/${arch}
fi
cp ${cache_dir}/${dist}.${ARCH}/* ${prefix}/var/cache/zypp/packages/opensuse/suse/${arch}


#
#  3.  Ensure that zypper has a working configuration file.
#
arch=i386
if [ $ARCH = "amd64" ] ; then
    arch=x86_64
fi

echo "  Creating zypper repo entry"
[ -d "${prefix}/etc/zypp/repos.d" ] || mkdir -p ${prefix}/etc/zypp/repos.d
cat > ${prefix}/etc/zypp/repos.d/${dist}.repo <<EOF
[opensuse]
name=${dist}
baseurl=$(dirname $(dirname ${mirror}))
enabled=1
gpgcheck=1

EOF


#
#  4.  Run "zypper install zypper".
#
echo "  Bootstrapping zypper"
chroot ${prefix} /sbin/ldconfig
chroot ${prefix} /usr/bin/zypper -n --no-gpg-checks install zypper      2>/dev/null
chroot ${prefix} /usr/bin/zypper -n --no-gpg-checks install vim 2>/dev/null
chroot ${prefix} /usr/bin/zypper -n --no-gpg-checks install dhcpcd    2>/dev/null
chroot ${prefix} /usr/bin/zypper -n --no-gpg-checks update              2>/dev/null


#
#  5.  Clean up
#
echo "  Cleaning up"
chroot ${prefix} /usr/bin/zypper clean
umount ${prefix}/proc
umount ${prefix}/sys


#
#  6.  Remove the .rpm files from the prefix root.
#
echo "  Final tidy..."
for i in ${prefix}/*.rpm; do
    rm -f $i
done
find ${prefix} -name '*.rpmorig' -delete
find ${prefix} -name '*.rpmnew' -delete


examples/rinse/usr/lib/rinse/prestage/opensuse-13.1.sh
#!/bin/bash
#

# author : Horst.venzke@remsnet.de
# License : LGPL v2
# (c) remsnet LTD
# https://github.com/remsnet/rinse


# there are better ways to to this  with groupadd / useradd

# rise with zypper ( opensuse based ) require  valid user base for the rpm installs
# without you get many errors with i.e " unknown user root / invalid user or group shadow ..."

# some uid / gids change with near all opensuse release
# any Hints are welcome.


prefix=$1


#
#  Source our common functions
#
if [ -e /usr/share/xen-tools/common.sh ]; then
    . /usr/share/xen-tools/common.sh
else
    . ./hooks/common.sh
fi


#
# Log our start
#
logMessage Script $0 starting


echo --------------------------------------
#
env
#
echo --------------------------------------
#
pwd
#
echo --------------------------------------
echo ${prefix}
echo --------------------------------------

#--------------opensuse aaa_base chroot hack--------------------------
#
# chroot basedirs
cd ${prefix}
mkdir -p -m 0755 etc
mkdir -p -m 0755 root
mkdir -p -m 1755 tmp
mkdir -p -m 0755 usr
mkdir -p -m 0755 home
mkdir -p -m 0755 bin
mkdir -p -m 0755 sbin
mkdir -p -m 0755 proc
mkdir -p -m 0755 media
mkdir -p -m 0755 lib
mkdir -p -m 0755 var/run
mkdir -p -m 0755 var/log
mkdir -p -m 0755 var/lock
mkdir -p -m 0755 var/tmp
mkdir -p -m 0755 var/adm
mkdir -p -m 0755 var/opt
mkdir -p -m 0755 var/spool
mkdir -p -m 0755 var/crash
mkdir -p -m 0755 var/tmp
mkdir -p -m 0755 usr/lib
mkdir -p -m 0755 usr/bin
mkdir -p -m 0755 usr/sbin
mkdir -p -m 0755 usr/share
mkdir -p -m 0755 usr/local
mkdir -p -m 0755 usr/include
mkdir -p -m 0755 usr/src
mkdir -p -m 0755 lib/udev/rules.d/


##We realy need  the base sample users - as YAST or autobase not used this way are _required_
# this partly emulate what yast installer do..

cd ${prefix}
# from opensuse 13.1 desktop
cat >>etc/passwd << EOF
root:x:0:0:root:/root:/bin/bash
lp:x:4:7:Printing daemon:/var/spool/lpd:/bin/bash
mail:x:8:12:Mailer daemon:/var/spool/clientmqueue:/bin/false
news:x:9:13:News system:/etc/news:/bin/bash
uucp:x:10:14:Unix-to-Unix CoPy system:/etc/uucp:/bin/bash
man:x:13:62:Manual pages viewer:/var/cache/man:/bin/bash
wwwrun:x:30:8:WWW daemon apache:/var/lib/wwwrun:/bin/bash
nobody:x:65534:65533:nobody:/var/lib/nobody:/bin/bash
EOF


cd ${prefix}
cat >>etc/group << EOF
root:x:0:
tty:x:5:
lp:x:7:
www:x:8:
mail:x:12:
news:x:13:
uucp:x:14:
shadow:x:15:
dialout:x:16:
audio:x:17:
utmp:x:22:
man:x:62:
nobody:x:65533:
EOF

cd ${prefix}
cat >>etc/shadow << EOF
root:!::::::
lp:*:16045::::::
mail:*:16045::::::
news:*:16045::::::
uucp:*:16045::::::
man:*:16045::::::
wwwrun:*:16045::::::
nobody:*:16045::::::
EOF

cd ${prefix}
cat >>etc/gshadow << EOF
abuild:*::
EOF

cd ${prefix}
cat >>etc/hosts << EOF
#               On small systems, this file can be used instead of a
#               "named" name server.
# Syntax:
#
# IP-Address  Full-Qualified-Hostname  Short-Hostname
#

127.0.0.1       localhost black.zq1.de

# special IPv6 addresses
::1             localhost ipv6-localhost ipv6-loopback

fe00::0         ipv6-localnet

ff00::0         ipv6-mcastprefix
ff02::1         ipv6-allnodes
ff02::2         ipv6-allrouters
ff02::3         ipv6-allhosts

EOF


### now install the chroot oss core with preveious created

# centos/fc/redhat rpm binary == /bin/rpm
cd ${prefix}


pwd
# cleanup rpm core  if exist
#
#---------------------------------------------------------
cd ${prefix}
test -d core || rm -rf ./core ; mkdir core
test -d core || rm -rf ./zypperrpm ; mkdir zypperrpm

# OpenSuSE 13.1 core - minimatistic chroot
# see https://github.com/remsnet/OpenSUSE-Bootstrap/blob/master/mk-mini-root.sh
#
echo "prepareing OSS 13.1 RPM rpm core"
cd ${prefix}
for i in aaa_base bash coreutils cpio cracklib cracklib diffutils filesystem fillup findutils glibc grep info insserv libacl1 libattr1 libaudit1 libblkid1 libbz2 libcap2 libcrack2 libdb libgcc_s1 libgdbm4 libgmp10 libgssglue1 liblzma5 libmount1 libncurses5 libpcre1 libreadline5 libreadline6 libselinux1 libstdc++6 libtirpc1 libutempter0 libuuid1 libz1 libzio1 ncurses openSUSE openSUSE pam pam perl perl permissions sed terminfo util insserv-compat
do
cd ${prefix}
/bin/cp --force $i-*.rpm core/
done

#
# zypper requirements
for y in libssh2
do
cd ${prefix}
/bin/cp --force $y-*.rpm zypperrpm/
done

/bin/rpm -i --root  ${prefix}/   ${prefix}/core/*.rpm
/bin/rpm -i --root  ${prefix}/   ${prefix}/zypperrpm/*.rpm
test -d core && rm -rf ./core
test -d zypperrpm && rm -rf ./zypperrpm

#--------------opensuse aaa_base  chroot prestage hack--end -----------------------


#
#  Log our finish
#
logMessage Script $0 finished
#
